lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,204 @@
+import json
+from http import HTTPStatus
+from logging import Logger
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    APIROLE,
+    CONTEXT,
+    ElasticBase,
+    JSON,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.database.models.tracker import Dataset, Metadata
+
+
+class DatasetsPublish(ElasticBase):
+    """
+    Change the "access" authorization of a Pbench dataset
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("controller", ParamType.STRING, required=True),
+                Parameter("name", ParamType.STRING, required=True),
+                Parameter("user", ParamType.USER, required=True),
+                Parameter("access", ParamType.ACCESS, required=True),
+            ),
+            role=APIROLE.UPDATE,
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this name, and determine whether the
+        authenticated user has UPDATE access to this dataset. (Currently, this
+        means the authenticated user is the owner of the dataset, or has ADMIN
+        role.)
+
+        Otherwise, record the dataset so that the postprocess operation can
+        mark it as published.
+
+        Args:
+            client_json: Request JSON payload
+
+        Raises:
+            Any errors in the postprocess method shall be reported by
+            exceptions which will be logged and will terminate the operation.
+
+        Returns:
+            dataset object if the operation should continue, or None
+        """
+        dataset = Dataset.attach(
+            controller=client_json["controller"], name=client_json["name"]
+        )
```

Comment:
Sadly, I realized belatedly that I omitted one security check, which goes here. We checked that the authorization header provides legitimate UPDATE access for data owned by the referenced user ... but we need to establish that the target dataset is actually owned by that user since we can't rely here on an Elasticsearch query term.